devicetypes/stelpro/stelpro-ki-thermostat.src/stelpro-ki-thermostat.groovy

```diff
@@ -0,0 +1,572 @@
+	log.warn "Unexpected zwave command $cmd"
+}
+
+def refresh() {
+	poll()
+}
+
+def configure() {
+	unschedule(scheduledUpdateWeather)
+	if (settings.zipcode) {
+		runEvery1Hour(scheduledUpdateWeather)
+	}
+	poll()
+}
+
+def quickSetHeat(degrees) {
+	setHeatingSetpoint(degrees, 0)
+}
+
+def setHeatingSetpoint(preciseDegrees, delay = 0) {
+	def degrees = new BigDecimal(preciseDegrees).setScale(1, BigDecimal.ROUND_HALF_UP)
+	log.trace "setHeatingSetpoint($degrees, $delay)"
+	def deviceScale = state.scale ?: 1
+	def deviceScaleString = deviceScale == 2 ? "C" : "F"
+	def locationScale = getTemperatureScale()
+	def p = (state.precision == null) ? 1 : state.precision
+
+	def convertedDegrees
+	if (locationScale == "C" && deviceScaleString == "F") {
+		convertedDegrees = celsiusToFahrenheit(degrees)
+	}
+	else if (locationScale == "F" && deviceScaleString == "C") {
+		convertedDegrees = fahrenheitToCelsius(degrees)
+	}
+	else {
+		convertedDegrees = degrees
+	}
+
+	sendEvent(name: "heatingSetpoint", value: degrees, unit: locationScale, data: [heatingSetpointRange: heatingSetpointRange])
+	sendEvent(name: "thermostatSetpoint", value: degrees, unit: locationScale, data: [thermostatSetpointRange: thermostatSetpointRange])
+
+	delayBetween([
+		zwave.thermostatSetpointV2.thermostatSetpointSet(setpointType: 1, scale: deviceScale, precision: p, scaledValue: convertedDegrees).format(),
+		zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1).format(),
+		poll()
+	], 1000)
+}
+
+def quickSetOutTemp(degrees) {
+	setOutdoorTemperature(degrees, 0)
+}
+
+def setOutdoorTemperature(degrees, delay = 0) {
+	setOutdoorTemperature(degrees.toDouble(), delay)
+}
+
+def setOutdoorTemperature(Double degrees, Integer delay = 0) {
+	def deviceScale
+	def locationScale = getTemperatureScale()
+	def p = (state.precision == null) ? 1 : state.precision
+	if (locationScale == "C") {
+		deviceScale = 0
+	}
+	else {
+		deviceScale = 1
+	}
+	log.debug "setOutdoorTemperature: ${degrees}"
+	zwave.sensorMultilevelV3.sensorMultilevelReport(sensorType: 1, scale: deviceScale, precision: p, scaledSensorValue: degrees).format()
+}
+
+def increaseHeatSetpoint() {
+	float currentSetpoint = device.currentValue("heatingSetpoint")
+	def locationScale = getTemperatureScale()
+	float maxSetpoint
+	float step
+
+	if (locationScale == "C") {
+		maxSetpoint = 30;
+		step = 0.5
+	}
+	else {
+		maxSetpoint = 86
+		step = 1
+	}
+
+	if (currentSetpoint < maxSetpoint) {
+		currentSetpoint = currentSetpoint + step
+		quickSetHeat(currentSetpoint)
+	}
+}
+
+def decreaseHeatSetpoint() {
+	float currentSetpoint = device.currentValue("heatingSetpoint")
+	def locationScale = getTemperatureScale()
+	float minSetpoint
+	float step
+
+	if (locationScale == "C") {
+		minSetpoint = 5;
+		step = 0.5
+	}
+	else {
+		minSetpoint = 41
+		step = 1
+	}
+
+	if (currentSetpoint > minSetpoint) {
+		currentSetpoint = currentSetpoint - step
+		quickSetHeat(currentSetpoint)
+	}
+}
+
+def switchMode() {
```

Comment:
since we're hardcoding the supported modes this should be a lot simpler of a function